### PR TITLE
[WIP] Document tests and try to isolate intermittent test failures

### PR DESCRIPTION
**Pull request by willingc**
- Adds pylint config to error on no docstrings
- As I have run the tests a bunch on travis, I've seen frequent, though intermittent failures, on a few of the tests (esp. `test_spawn_form` and `test_groups`). It looks like the failures are when there is no connection to the socket and a socket.gaierror is thrown. `socket.getsockname` is involved.
- The errors seem to happen more frequently when testing the subdomain environment variable.

Going to continue to document and trace through the code base while running tests to see where it leads.

### CODE REVIEW

File: jupyterhub/tests/conftest.py [hunk 1]
```diff
@@ -43,73 +46,85 @@ def db():
...
+    """Create a Tornado IOLoop and make it current."""
     loop = ioloop.IOLoop()
     loop.make_current()
     return loop
 
 
 @fixture(scope='module')
 def app(request):
+    """Start a Hub"""
+    # TODO - Q: what tracks the MockHub instance so it can be cleared later.
```
Comment (minrk) [inline jupyterhub/tests/conftest.py]: The `.instance` method itself does this. It's tracked on the class. The instance is cleared with `clear_instance` in the finalizer immediately below.

File: jupyterhub/utils.py [hunk 2]
```diff
@@ -104,7 +104,9 @@ def auth_decorator(check_auth):
     in your decorator, so you can decorate while you decorate.
     """
     def decorator(method):
+        """Make a method decorator."""
         def decorated(self, *args, **kwargs):
+            """Make a decorator to check authentication."""
```
Comment (minrk) [inline jupyterhub/utils.py]: `decorated` should copy `method`'s docstring, not add its own.

File: .pylintrc [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+[REPORTS]
+# Do not print a summary report
+reports=no
+
+[FORMAT]
+# One of the few stylistic things we try to check.
+# Modules shouldn't be *too large* - should be broken up
+# Feel free to bump this number if you disagree
+max-module-lines=2000
```
Comment (minrk) [inline .pylintrc]: I don't think we should check module length at all.

File: jupyterhub/tests/test_orm.py [hunk 4]
```diff
@@ -108,47 +111,56 @@ def test_tokens(db):
...
     assert service
     assert server.id is None
+
     db.commit()
     assert isinstance(server.id, int)
-    
+
 
 def test_token_find(db):
+    """Test if token is in db"""
```
Comment (minrk) [inline jupyterhub/tests/test_orm.py]: This would more accurately be "Test APIToken.find"

File: jupyterhub/tests/conftest.py (hunk 1 above)
Comment (willingc) [inline jupyterhub/tests/conftest.py]: @minrk Thanks! If it's cool with you, I will likely flag questions like this on WIP PRs when I have specific code questions.

Also, you might find this article by Miguel Grinberg interesting: https://blog.miguelgrinberg.com/post/unit-testing-asyncio-code I haven't dug into it yet but it likely contains some good nuggets.

Comment (minrk) [inline jupyterhub/tests/conftest.py]: Yup, I think that's quite sensible. It comes up at the right time during review.

Comment (minrk) [inline jupyterhub/tests/conftest.py]: Also, I saw that fly by on twitter, and indeed it looks super useful, thanks!

**Comment by willingc** [comment]
- [ ] conftest.py
- [ ] mocking.py
- [ ] mockservice.py
- [ ] mocksu.py
- [ ] test_api.py
- [ ] test_app.py
- [ ] test_auth.py
- [ ] test_db.py
- [ ] test_orm.py
- [ ] test_pages.py
- [ ] test_proxy.py
- [ ] test_services.py
- [ ] test_services_auth.py
- [ ] test_singleuser.py
- [ ] test_spawner.py
- [ ] test_traitlets.py
- [ ] test_utils.py

**Comment by willingc** [comment]
Note to self re: segfaults in travis: Investigate MockHub in mocking.py. Check logic of `evt = threading.Event()` and `self._thread = threading.Thread(target=_start)` on starts, stops, waits, and timeouts. May be missing a guard.

**Comment by willingc** [comment]
Pretty confident that #984 finds the seg fault error with the pytest tests added by @minrk. See [travis output](https://travis-ci.org/jupyterhub/jupyterhub/jobs/201193123). Let's try to get #984 merged :-)

**Comment by willingc** [comment]
Closing this stale PR since #1281 resolves.

**Pull request closed without merging**